### PR TITLE
[LG] Make template name format consistent with function name format

### PR DESCRIPTION
**Pull request by Danieladu**
close: #3942

The function name's format cannot cover all template name's formats. 

The valid format of template name is 
```
start with letter/number/_ and can only contains letter/number/./_
```
And the valid format of function name should meet
```
foreach(var part in expression.split('.')) 
{
    // part should start with letter/_ and can only contains letter/number/_
}
```

So, in some cases, user use template evaluation in expression with the "gap" format would throw exception.

Here is the remedy that disables the name that begins with number in each part of the split result from dot(.).

Now, the valid format of template name should be :
```
foreach(var part in expression.split('.')) 
{
    // part should start with letter/_ and can only contains letter/number/_
}
```
Same with the expression function name.

### CODE REVIEW

**Comment by boydc2014** [comment]
If i understand correctly, our allowed template name format is a little bit more than allowed function name.  And the proposed fix here is to restrict the template name format to fix the gap. 

My question is why not extending the allowed function name to reach the scope of allowed template name? 

One advice is maybe list the rules for template name and function name using just Regex pattern will helps clarify the issues, the description above is not very clear.

**Comment by Danieladu** [comment]
> If i understand correctly, our allowed template name format is a little bit more than allowed function name. And the proposed fix here is to restrict the template name format to fix the gap.
> 
> My question is why not extending the allowed function name to reach the scope of allowed template name?
> 
> One advice is maybe list the rules for template name and function name using just Regex pattern will helps clarify the issues, the description above is not very clear.

In other programming languages, it is often not allowed to use numbers at the beginning of the function name, and it is not possible to directly use numbers as function names. 
But I have no enough reason why not extending the allowed function name to reach the scope of allowed template name because from the current architecture, we do support this.
@vishwacsena have any ideas about this one?

**Comment by boydc2014** [comment]
> If i understand correctly, our allowed template name format is a little bit more than allowed function name. And the proposed fix here is to restrict the template name format to fix the gap.
> 
> My question is why not extending the allowed function name to reach the scope of allowed template name?
> 
> One advice is maybe list the rules for template name and function name using just Regex pattern will helps clarify the issues, the description above is not very clear to me.

After thinking about that, we should not allow number as start of an idenfiter (function name or template name, or part of function name and template name). It will give `12` as a validate function name or variable name, totally confusing the parser\evaluator.   More context is to https://stackoverflow.com/questions/342152/why-cant-variable-names-start-with-numbers

I agree with the downsize the scope of allowed template to the scope of allowed function name.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateErrors.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public class TemplateErrors
     {
         public const string NoTemplate = "LG file must have at least one template definition.";
 
-        public const string InvalidTemplateName = "Invalid template name. Template name should start with letter/number/_ and can only contains letter/number/./_.";
+        public const string InvalidTemplateName = "Invalid template name. Each part of template name split by dot(.) should start with letter/_ and can only contain letter/_/number.";
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateErrors.cs:13]: contain\contains

**Comment by vishwacsena** [comment]
So what exactly are we going to disallow with template names? That they cannot start with a number? Can anyone ever have a template that starts with a number in the GA packages that will be broken by this change?

**Comment by Danieladu** [comment]
> So what exactly are we going to disallow with template names? That they cannot start with a number? Can anyone ever have a template that starts with a number in the GA packages that will be broken by this change?

Yes. If some one use template name with number start, and evaluate it in other template, for example, 
```
# 1
- template 1

# template
- ${1()} // this would be broken, and throw the exception: the expression is not invalid.
```

this pr is make the restrict from template name directly, finally, would get:
```
# 1 // template name is invalid, please use ****
- template 1

# template
- ${1()}
```

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateErrors.cs (hunk 1 above)
Comment (vishwacsena) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateErrors.cs:13]: Invalid template name. Template names can only contain letter, underscore '_' or a number. Any part of a template name (split by '.') cannot start with a number.

**Pull request merged**